src/posix/platform/radio_url.cpp

```diff
@@ -0,0 +1,224 @@
+    " spi-cs-delay[=usec]       Specify the delay after C̅S̅ assertion, in µsec.\n"                  \
+    " spi-reset-delay[=ms]      Specify the delay after R̅E̅S̅E̅T̅ assertion, in milliseconds.\n"  \
+    " spi-align-allowance[=n]   Specify the maximum number of 0xFF bytes to clip from start of\n"       \
+    "                           MISO frame. Max value is 16.\n"                                         \
+    " spi-small-packet=[n]      Specify the smallest packet we can receive in a single transaction.\n"  \
+    "                           (larger packets will require two transactions). Default value is 32.\n"
+#else
+#define RADIO_URL_HELP_BUS                                                                      \
+    "Radio URL:\n"                                                                              \
+    "spinel+hdlc+uart://${PATH_TO_UART_DEVICE}?${arguments} for real uart device\n"             \
+    "spinel+hdlc+fortpty://${PATH_TO_UART_DEVICE}?${arguments} for forking a pty subprocess.\n" \
+    "arguments:\n"                                                                              \
+    " uart-parity[=even|odd]     Uart parity config, optional.\n"                               \
+    " uart-stop[=number-of-bits] Uart stop bit, default is 1.\n"                                \
+    " baudrate[=baudrate]        Uart baud rate, default is 115200.\n"                          \
+    " arg[=argument string]      Command line arguments for subprocess, can be repeated.\n"
+#endif
+    return RADIO_URL_HELP_BUS " no-reset                   Skip resetting the radio device.\n"
+                              " ncp-dataset                Retrieve dataset from ncp.\n"
+                              " max-power-table            Max power table for each channel, splitted by comma.\n";
+}
+
+namespace ot {
+namespace Posix {
+
+Arguments::Arguments(const char *aUrl)
+{
+    char *url = &mUrl[0];
+
+    mPath  = NULL;
+    mStart = NULL;
+    mEnd   = NULL;
+
+    VerifyOrExit(aUrl != NULL, OT_NOOP);
+    VerifyOrExit(strnlen(aUrl, sizeof(mUrl)) < sizeof(mUrl), OT_NOOP);
+    strncpy(mUrl, aUrl, sizeof(mUrl) - 1);
+    url = strstr(url, "://");
+    VerifyOrExit(url != NULL, OT_NOOP);
+    url += sizeof("://") - 1;
+    mPath = url;
+
+    mStart = strstr(url, "?");
+
+    if (mStart != NULL)
+    {
+        mStart[0] = '\0';
+        mStart++;
+
+        mEnd = mStart + strlen(mStart);
+    }
+    else
+    {
+        mEnd = url + strlen(url);
+    }
+exit:
+    return;
+}
+
+const char *Arguments::GetValue(const char *aName, const char *aLastValue)
+{
+    const char * rval  = NULL;
+    const size_t len   = strlen(aName);
+    char *       start = (aLastValue == NULL ? mStart : (const_cast<char *>(aLastValue) + strlen(aLastValue) + 1));
+
+    while (start < mEnd)
+    {
+        char *last = NULL;
+
+        for (char *cur = strtok(start, "&"); cur != NULL; cur = strtok(NULL, "&"))
```

Comment:
I assume we should be able to call `GetValue()` with different `aName` sequentially and in any order we want. Say  we have a URL string such that `mStart` is `"first=1&second=2&third=3"`, we should be able to call `GetValue("third")` first and then `GetValue("second")`.

If I am not mistaken, with current code this would not work.
- The `strtok` would actually update the original string itself and replace the delimiter `&` with `\0`
- So I think first time going through this, we would get all tokens but now the original string does not have `"&"` anymore between each toekn/arg
  - The original string pointed by `mStart` would become `"first=1\0second=2\0third=3"`
- Next call to `GetValue()` we again start from `mStart` but this time it cannot go through the full list since now `strtok` would only see `"first=1"`

Reply:
I re-read the code and realized how it handles the changed the `mStart` string (by remembering `last` string and skipping over it and the `\0` from the `while (start < mEnd)` loop). So the code should work correctly as is.

How about we do the `strtok` from the constructor (which then replaces all `&` with `\0` in string once) and then in `GetValue()` just use the iteration from `start` to `mEnd` skipping over each string by its length? I feel this helps make the code simpler and avoid confusion (like my mistake reading it last night... :)...).